### PR TITLE
gRPC service for kv storage

### PR DESCRIPTION
**Pull request by kirugan**
Issue: #743

### CODE REVIEW

File: grpc/handlers.go [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+			return err
+		}
+	}
+}
+
+func (h handlers) handleTxCursor(
+	cur *gen.Cursor,
+	tx *tx,
+	server gen.KV_TxServer,
+) error {
```
Comment (rianhughes) [inline grpc/handlers.go:36]:
```suggestion
func (h handlers) handleTxCursor(cur *gen.Cursor, tx *tx, server gen.KV_TxServer) error {
```

File: grpc/handlers.go [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+	if err != nil {
+		return err
+	}
+
+	switch cur.Op {
+	case gen.Op_SEEK:
+		key := utils.Flatten(cur.BucketName, cur.K)
+		it.Seek(key)
+	case gen.Op_NEXT:
+		it.Next()
```
Comment (rianhughes) [inline grpc/handlers.go:47]: Just a question: how are `it.Next()` and `it.Seek(key)` being used here? Do you not need to return the results to the caller?

Comment (kirugan) [reply grpc/handlers.go:47]: For both methods return value is `bool` and I tried to follow existing proto spec (there is no field that corresponds to wrong iterator). Probably in this case it's better to return error, will check

File: grpc/handlers.go [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+)
+
+type handlers struct {
+	db db.DB
+}
+
+func (h handlers) Tx(server gen.KV_TxServer) error {
+	dbTx := h.db.NewTransaction(false)
+	tx := newTx(dbTx)
+	defer tx.cleanup()
```
Comment (omerfirmak) [inline grpc/handlers.go:17]: As a rule, we don't defer functions that can error.

See `db.CloseAndWrapOnError` and please use that to explicitly call cleanup on return statements.

File: grpc/tx.go [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+		t.iterators[id] = it
+	}
+
+	return it, nil
+}
+
+func (t *tx) cleanup() error {
+	for _, it := range t.iterators {
+		it.Close()
```
[the remark this answers is not in the document]
Comment (kirugan) [reply grpc/tx.go]: I updated go to 1.20 in `go.mod` and `README.md` (we already use it in Docker). There is handy function for that - `errors.Join(...)` ❤️

File: grpc/handlers.go (hunk 3 above)
Comment (kirugan) [reply grpc/handlers.go]: Rewrote it with `CloseAndWrap...` function.

File: grpc/grpc_test.go [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+		Op: gen.Op_CURRENT,
+	})
+	require.NoError(t, err)
+
+	pair, err := stream.Recv()
+	if err != nil {
+		if err == io.EOF {
+			fmt.Println("disconnected from server")
+		} else {
+			spew.Dump("error", err)
```
Comment (omerfirmak) [inline grpc/grpc_test.go:42]: I think require.NoError does this already, no?

File: node/node.go [hunk 6]
```diff
@@ -149,6 +151,10 @@ func makeHTTP(port uint16, rpcHandler *rpc.Handler, log utils.SimpleLogger) *jso
 	}, log)
 }
 
+func makeGRPC(port uint16, db db.DB, log utils.Logger) *grpc.Server {
+	return grpc.NewServer(port, db, log)
+}
+
 // Run starts Juno node by opening the DB, initialising services.
```
Comment (omerfirmak) [inline node/node.go:158]: we can inline this?

**Review by omerfirmak — APPROVED** [review]
I dont think we need to delay this any longer, let's include it in a release and allow apibara to start testing it.

File: grpc/grpc_test.go (hunk 5 above)
Comment (kirugan) [reply grpc/grpc_test.go:42]: `spew.Dump(...)` is very convinient pretty printer in Go, I was debugging what kind of grpc errors grpc can I get

File: node/node.go (hunk 6 above)
Comment (kirugan) [reply node/node.go]: sure, I just wanted to mimic http service with its `makeHTTP(...)`. inlined

File: grpc/handlers.go (hunk 2 above)
Comment (joshklop) [reply grpc/handlers.go:47]: Was reading through the code and almost posted the same question :)

Is it better to return the error here @kirugan?

**Review by joshklop — APPROVED** [review]

**Comment by joshklop** [comment]
Another thought: unit tests for the grpc package could be a good-first-issue.

File: grpc/handlers.go (hunk 2 above)
Comment (kirugan) [reply grpc/handlers.go:47]: Ok, I'll return error here. The problem with provided proto file is that there is no error => cursor correspondence, probably Rust code will add it later.

**Comment by IronGauntlets** [comment]
> Another thought: unit tests for the grpc package could be a good-first-issue.

In general, we don't want this to become a practice. I think it is important for the author of the PR to include tests in the PR. I don't think it is fair on others to write tests for code that was authored by someone else as very few people look forward to writing tests. 

We have `codecov/patch` to keep track of this and ideally, we should aim for 80% (it's not always possible that's why it is not a required condition for merging). 1.08% coverage doesn't seem enough.

For now, we can merge it but @kirugan should write tests for it soon.

File: grpc/handlers.go [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+		if err != nil {
+			return err
+		}
+
+		err = server.Send(&gen.Pair{
+			K:        it.Key(),
+			V:        v,
+			CursorId: cur.Cursor,
+		})
+	default:
```
Comment (fracek) [inline grpc/handlers.go:60]: I believe `Op_FIRST` is not handled.

File: grpc/handlers.go (hunk 2 above)
Comment (fracek) [reply grpc/handlers.go:47]: IIRC in rust in case of error we get an empty pair response (with only the cursor id set).

**Comment by fracek** [comment]
First of all, thank you for this! I played around with this and here's my thoughts:

 * It would be nice to have a `Version()` method that returns the current Juno version. Not because it's _that_ useful to know the version, but because it makes it straight forward to test if the connection is going to work (easier than setting up bidi stream).
 * It looks like the `Open` operation is not supported? It's useful to get a `cursor_id` assigned by the server.
 * Since there are no buckets, is it possible to have a `seek_range` operator that positions at the first element greater than or equal the provided key? That way `seek_range(table_prefix || 0)` is equivalent to positioning to the first element of the table.
   - If that's the behavior of `seek`, then it would be useful to have a `seek_exact` that returns an empty kv if the key doesn't match exactly.
 * How would I implement something like `last`? 
 * I also added some smaller comments to the discussion.

**Comment by kirugan** [comment]
@fracek 

> It would be nice to have a Version() method ...

Implemented 

> It looks like the Open operation is not supported? It's useful to get a cursor_id assigned by the server.

Initially I thought client should send `cursorId` and if doesn't exist server creates new iterator for new `cursorId`. I've changed this logic, added `open` operation

> If that's the behavior of seek, then it would be useful to have a seek_exact that returns an empty kv if the key doesn't match exactly.

yep, current `seek` works the way you described `seek_range` operation. I've also added `seek_exact`

> How would I implement something like last?

The only way to do it right now is to iterate over all keys with same prefix until you reach next prefix key. I know it's not super convenient, I'll think how to fix it.

> I also added some smaller comments to the discussion.

Thanks

File: grpc/handlers.go (hunk 7 above)
Comment (kirugan) [reply grpc/handlers.go:99]: I thought it's the same as `Op_SEEK` + `Op_Current`

**Comment by kirugan** [comment]
I'll merge this PR. The main functionality is implemented. All subsequent edits and features will be added to other pr

**Pull request merged**